csrc/device_lower/pass/allocation.cpp

```diff
@@ -551,25 +564,80 @@ class AllocationInserter : public kir::ExprMutator {
-                                 .shape(std::vector<int64_t>{})
-                                 .dtype(DataType::UInt)
-                                 .contiguity(true)
-                                 .build();
-      mbarrier->setMemoryType(MemoryType::Shared);
-      auto mbarrier_init = IrBuilder::create<kir::MBarrierInit>(
-          mbarrier, expr->container()->oneVal(DataType::UInt32));
-      auto mbarrier_inval =
-          IrBuilder::create<kir::MBarrierInvalidate>(mbarrier);
-
-      kir::Allocate* mbarrier_alloc =
-          IrBuilder::create<kir::Allocate>(mbarrier, MemoryType::Shared);
-      kir::Scope* expr_scope = scope_.empty() ? nullptr : scope_.back();
-      registerInsertBefore(expr, mbarrier_alloc, expr_scope);
-      registerInsertBefore(expr, mbarrier_init, expr_scope);
-      registerInsertAfter(expr, mbarrier_inval, expr_scope);
-      GpuLower::current()->ldstMBarrierMap()[expr] = mbarrier;
+      if (circular_buffer_depth > 1) {
+        // Create and allocate a memory barrier, if this is a circular buffer
+        //  then allocate an array of mbarier objects.
+        // mbarrier::init and mbarrier::inval will be updated in Double
+        //  Buffering pass, but we need to add them to ensure smem placeholders
+        //  lifetime in alias memory pass.
+        TensorView* mbarrier =
+            TensorViewBuilder()
+                .shape(std::vector<int64_t>{circular_buffer_depth})
+                .dtype(DataType::UInt)
+                .contiguity(true)
+                .build();
+        mbarrier->setMemoryType(MemoryType::Shared);
+
+        kir::Allocate* mbarrier_alloc =
+            IrBuilder::create<kir::Allocate>(mbarrier, MemoryType::Shared);
+        kir::Scope* expr_scope = scope_.empty() ? nullptr : scope_.back();
+
+        auto mbarrier_init = IrBuilder::create<kir::MBarrierInit>(
+            mbarrier, expr->container()->oneVal(DataType::UInt32));
+        auto mbarrier_inval =
+            IrBuilder::create<kir::MBarrierInvalidate>(mbarrier);
+
+        // For double/circular buffers we need to prepare a placeholder
+        //  for sync tokens created by 'MBarrierArriveExpectTx' IR node.
+        // Tokens are placed in smem and used by threads in a block.
+        TensorView* mbarrier_tokens =
+            TensorViewBuilder()
+                .shape(std::vector<int64_t>{circular_buffer_depth})
+                .dtype(DataType::UInt)
+                .contiguity(true)
+                .build();
+        mbarrier_tokens->setMemoryType(MemoryType::Shared);
+        kir::Allocate* mbarrier_tokens_alloc = IrBuilder::create<kir::Allocate>(
+            mbarrier_tokens, MemoryType::Shared);
+
+        registerInsertBefore(expr, mbarrier_tokens_alloc, expr_scope);
+        registerInsertBefore(expr, mbarrier_alloc, expr_scope);
+        registerInsertBefore(expr, mbarrier_init, expr_scope);
+        registerInsertAfter(expr, mbarrier_inval, expr_scope);
```

Comment:
So we are doing the init and inval inside the loop of cp.async, i.e. generate code like:
```
for i in range(I0):
    __shared__ tokens;
    __shared__ mbarrier;
    init(mbarrier);
    cp.async.bulk(data, mbarrier);
    inval(mbarrier);
```
and leave it for the double bufferring pass to fix it? Could you leave a comment in the code for this?

Reply:
Yes, that's the current implementation - I will update the code here so it is clear that Double Buffering pass modify this part of the code. Having it done here has the benefit of easier control of shared memory allocations / storage liveness (the order of passes is `insertAllocations`-> `reuseMemoryAllocations` -> `DoubleBufferPass`).

Switching to approach, where all needed elements are added in double buffering pass (we detect `LoadStoreOp` for bulk copy op -> we add placeholder for tokes, mbarriers, init mbarrier and later inval for mbarriers) will have major consequences: adding to the system smem allocation (mbarrier / tokens) and life time analysis out-of-the order (pass-wise) might be a challenge.

I will think how we could have this done, in a clean way.

Thank you for the comment!